File: extensions/oh-my-gemini/scripts/session_end.py
#!/usr/bin/env python3
import json
import os
import re
import sys
from pathlib import Path


def check_task_completed(task_path: Path) -> bool:
    task_plan_file = task_path / "task_plan.md"
    if not task_plan_file.exists():
        return False

    try:
        content = task_plan_file.read_text()
        status_pattern = r"- \*\*Status:\*\* (\w+)"
        statuses = re.findall(status_pattern, content)

        if not statuses:
            return False

        return all(s == "complete" for s in statuses)

    except Exception:
        return False


def main():
    project_workspace_path = os.environ.get("WORKSPACE_PATH", "")

    try:
        input_data = json.load(sys.stdin)
    except:
        input_data = {}

    if project_workspace_path:
        working_dir_base = Path(project_workspace_path) / ".agent_working_dir"
    else:
        working_dir_base = Path(os.getcwd()) / ".agent_working_dir"

    current_task_file = working_dir_base / "current_task.json"

    if not current_task_file.exists():
        print(json.dumps({"decision": "allow"}), flush=True)
        return

    try:
        with open(current_task_file, "r") as f:
            current_task_data = json.load(f)

        current_task = current_task_data.get("current")

        if not current_task:
            print(json.dumps({"decision": "allow"}), flush=True)
            return

        task_path = Path(current_task)

        if check_task_completed(task_path):
            current_task_data["current"] = None
            with open(current_task_file, "w") as f:
                json.dump(current_task_data, f, indent=2)

            task_name = task_path.name.replace("task_", "").replace("_", " ")

            print(
                json.dumps(
                    {
                        "systemMessage": f"[oh-my-gemini] Task '{task_name}' completed and reset."
                    }
                ),
                flush=True,
            )
        else:
            task_name = task_path.name.replace("task_", "").replace("_", " ")

            print(
                json.dumps(
                    {
                        "systemMessage": f"[oh-my-gemini] Session ended. Task '{task_name}' saved. Run /omg:resume to continue."
                    }
                ),
                flush=True,
            )

    except (json.JSONDecodeError, IOError):
        pass

    print(json.dumps({"decision": "allow"}), flush=True)


if __name__ == "__main__":
    main()

File: extensions/oh-my-gemini/scripts/session_end.js
#!/usr/bin/env node
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * SessionEnd Hook
 * 
 * Checks task completion status on session end.
 * - If task is complete, reset current task
 * - If task is incomplete, notify user (silently - user is exiting)
 * 
 * Note: Always exits with code 0 to avoid blocking session end
 */

const fs = require('fs');
const path = require('path');

const WORKSPACE_PATH = process.env.WORKSPACE_PATH || '';
const workingDirBase = path.join(WORKSPACE_PATH, '.agent_working_dir');
const currentTaskFile = path.join(workingDirBase, 'current_task.json');

/**
 * Check if all phases in task_plan.md are complete
 * @param {string} taskPath - Path to task directory
 * @returns {boolean} - True if all phases are complete
 */
function checkTaskCompleted(taskPath) {
  const taskPlanFile = path.join(taskPath, 'task_plan.md');
  
  if (!fs.existsSync(taskPlanFile)) {
    return false;
  }

  try {
    const content = fs.readFileSync(taskPlanFile, 'utf-8');
    const statusPattern = /-\s*\*\*Status:\*\*\s*(\w+)/g;
    const statuses = [];
    let match;
    
    while ((match = statusPattern.exec(content)) !== null) {
      statuses.push(match[1]);
    }

    if (statuses.length === 0) {
      return false;
    }

    return statuses.every(s => s === 'complete');
  } catch (err) {
    return false;
  }
}

/**
 * Output JSON to stdout
 * @param {object} data - Data to output
 */
function outputJson(data) {
  console.log(JSON.stringify(data));
}

function main() {
  // Read stdin
  let input_data = {};
  try {
    const stdin = fs.readFileSync(0, 'utf-8');
    input_data = JSON.parse(stdin);
  } catch (err) {
    // Ignore parsing errors
  }

  if (!fs.existsSync(currentTaskFile)) {
    outputJson({ decision: 'allow' });
    return;
  }

  try {
    const currentTaskData = JSON.parse(fs.readFileSync(currentTaskFile, 'utf-8'));
    const currentTask = currentTaskData.current;

    if (!currentTask) {
      outputJson({ decision: 'allow' });
      return;
    }

    if (checkTaskCompleted(currentTask)) {
      // Task is complete, reset
      currentTaskData.current = null;
      fs.writeFileSync(currentTaskFile, JSON.stringify(currentTaskData, null, 2));
    }
    // Task is incomplete - do nothing, user is exiting
  } catch (err) {
    // Non-fatal error
  }

  outputJson({ decision: 'allow' });
}

main();
